### ISSUE TITLE
Class netrefs return wrong __class__ value

### ISSUE THREAD

**Issue by jtobiska** [tomerfiliba/rpyc#355]
As of release 4.1, it seems that netrefs that represent classes are returning the wrong values for the `__class__` attribute.

```python
# float.__class__ from rpyc:
>>> import rpyc
>>> conn = rpyc.classic.connect("localhost")
>>> conn.modules.__builtin__.float.__class__
<type 'float'>

# float.__class__ locally from python
>>> float.__class__
<type 'type'>
```

We were formerly using version 3.3.0, which didn't have this issue. 

To complicate matters, it looks like the fix for #339 included a change to the implementation of `BaseNetref.__instancecheck__` that is now relying on this incorrect behavior.  The last line of `__instancecheck__` is now calling `isinstance(other, self.__class__)` rather than `isinstance(other, self)`.  `self` represents the correct class to check, but `self.__class__` is working because `self.__class__` is incorrectly returning `self` (or the local value of `self`, depending on whether the class is a builtin).

##### Environment

- rpyc version: 4.1.2
- python version: 2.7.13
- operating system: linux

**Comment by jtobiska** [comment]
It looks like the source of the problem is in `rpyc.core.protocol.Connection._netref_factory`.  In that function, `id_pack[2]` isn't being checked to see if the value is 0 to know if `id_pack` is referring to a type or an instance.  Instead the function seems to be always assuming it's an instance, so when `id_pack[0]` finds a match in `netref.builtin_classes_cache` it constructs and instance of that type rather than that type's type as it should.

**Comment by comrumino** [comment]
For brevity, I'm going to focus on ideal behavior compared to current behavior.

`{•̃_•̃}`

[**Definitions**](https://rpyc.readthedocs.io/en/latest/docs/theory.html#theory):
* Transferred by-value, serialization and transmission of an immutable object
* Referencing a remote object, using the unboxed reference to a mutable object
* Object proxying, the process of using a local object to forward behavior over a network connection to an object in another address space
* Netref, shorthand for network reference and a proxy object
* Data descriptor, settle for a [descriptor example](https://docs.python.org/3/howto/descriptor.html?highlight=invoking%20descriptors#descriptor-example)    

Of course, these definitions are specific to the context of RPyC

**Steps to reproduce with Python3:**
```
pkgver=<desired-version>
myvenv=rpyc${pkgver}
virtualenv /tmp/$myvenv
source /tmp/$myvenv/bin/activate
pip install rpyc==$pkgver
curl https://raw.githubusercontent.com/tomerfiliba/rpyc/$pkgver/bin/rpyc_classic.py -o /tmp/$myvenv/bin/rpyc_classic.py
python /tmp/$myvenv/bin/rpyc_classic.py &> /tmp/${myvenv}-rpyc_classic.log & disown
sleep 1
python -c 'import rpyc; conn = rpyc.classic.connect("127.0.0.1"); print("repr(conn.modules.builtins.float) prints as ", repr(conn.modules.builtins.float), "  #"); print("type(conn.modules.builtins.float) prints as ", type(conn.modules.builtins.float), "  #"); print("repr(conn.modules.builtins.float.__class__) prints as ", repr(conn.modules.builtins.float.__class__), "  #"); print("type(conn.modules.builtins.float.__class__) prints as ", type(conn.modules.builtins.float.__class__), "  #");'
deactivate
```

Running with `pkgver=4.1.2` and filling in the comments:
```
repr(conn.modules.builtins.float) prints as  <class 'float'>   # looks good
type(conn.modules.builtins.float) prints as  <netref class 'rpyc.core.netref.builtins.float'>   # looks good
repr(conn.modules.builtins.float.__class__) prints as  <class 'float'>   # looks good
type(conn.modules.builtins.float.__class__) prints as  <class 'type'>   # faint smell of a dumpster
```

The most intuitive behavior for the last line:
```
type(conn.modules.builtins.float.__class__) prints as  <netref class 'rpyc.core.netref.builtins.type'>
```
This is where transparency and secure default behavior collide. The special attribute `__class__` is not always safe. In our case, creating a netref to `builtins.type` provides access to a dangerous object (built-ins could create arbitrary code).

So, an ideal solution might be:
* add exception for `UnsafeAttributeUnavailable`
* `raise UnsafeAttributeUnavailable` when applicable
* create a data descriptor for unsafe special attributes (i.e. `__class__`)

**tl;dr** versions `<=4.0.2` should have provided a proxy to the float class. The current behavior is still not ideal, but I would like to drop Python2 support before fixing to reduce complexity.

**Comment by jtobiska** [comment]
Thanks for the detailed response to this issue.  The results from the examples you listed are still problematic though, regardless of whether it is netrefs or local objects that are returned.  The fundamental problem is that in Python ```type(float)``` is not ```float```, and ```float.__class__``` is not ```float```.  ```type(float)``` is ```type``` and ```float.__class__``` is ```type```.  It is critical that rpyc reflects those same results, regardless of whether it returns those results as netrefs or local objects.

**Comment by comrumino** [comment]
I hope this is agreeable.

No worries, the next release will support your use but unfortunately will still be backwards incompatible. If a [CVE](https://cve.mitre.org/cgi-bin/cvename.cgi?name=CVE-2019-16328) is okay then upgrading to version `4.0.2` will work meanwhile. 


**Supporting your use:** 
Suppose I changed
```
type(conn.modules.builtins.float.__class__) prints as  <class 'type'>
```
to the more intuitive behavior
```
type(conn.modules.builtins.float.__class__) prints as  <netref class 'rpyc.core.netref.builtins.type'>
```
when `allow_all_attrs` is specified your use case would be supported, but still backwards incompatible. The benefits of this approach:
1. Secure by default (in the future, support ease of extension to restrict behavior)
2. Supports your use case

**The reasoning behind the change in behavior**

Checking out #316 might provide some insight in the changes. 

If `conn.modules.builtins.float` is a `netref` to `float`, then `conn.modules.builtins.float` is a proxy object to a `float`. So, `type(conn.modules.builtins.float) prints as  <netref class 'rpyc.core.netref.builtins.float'>` is correct. 

Copying `test_magic.py` and `test_threaded_server.py` from `master` and executing with `rpyc==4.0.2` 
```
(venvrpyc2) [archie@swanstation tmp]$ python -m unittest test_threaded_server 
...
======================================================================
FAIL: test_instancecheck_across_connections (test_threaded_server.Test_ThreadPoolServer)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "test_threaded_server.py", line 37, in test_instancecheck_across_connections
    self.assertFalse(isinstance(bar, conn.modules.test_magic.Foo))
AssertionError: True is not false

======================================================================
FAIL: test_instancecheck_across_connections (test_threaded_server.Test_ThreadedServer)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "test_threaded_server.py", line 37, in test_instancecheck_across_connections
    self.assertFalse(isinstance(bar, conn.modules.test_magic.Foo))
AssertionError: True is not false

----------------------------------------------------------------------
Ran 5 tests in 4.619s

FAILED (failures=2)
```
when executing with `rpyc==4.1.2` the result
```
(venvrpyc2) [archie@swanstation tmp]$ python -m unittest test_threaded_server 
...
----------------------------------------------------------------------
Ran 5 tests in 4.420s

OK
```
If we look at the actual code from this test
```
        foo = conn.modules.test_magic.Foo()
        bar = conn.modules.test_magic.Bar()
        self.assertTrue(isinstance(foo, conn.modules.test_magic.Foo))
        self.assertTrue(isinstance(bar, conn2.modules.test_magic.Bar))
        self.assertFalse(isinstance(bar, conn.modules.test_magic.Foo))
```
Hence, the change in behavior.

**Comment by comrumino** [comment]
Note, this is my top priority issue. I'll see if I can get a fix within the week and thanks for reporting it :1st_place_medal:

I apologize if my language is brusque and I appreciate your feedback. Still working on the finer details for my mental model of RPyC (as a new-ish maintainer)---I get the issue now :100:

**Comment by jtobiska** [comment]
Thanks so much!  I appreciate all your help and your detailed responses.  Will the fix work for Python 2.7 or will it be Python 3 only?

**Comment by comrumino** [comment]
I think master should have a fix for this issue for both Python 2.7, 3.6, 3.7, and 3.8-dev. I did add a unit test under `test_get_id_pack.Test_get_id_pack.test_netref` based on the issue as described. I'll wait to close the issue in hopes that you can confirm it is fixed :100:  (the Travis failure for 3.8 is unrelated to this issue)

**Comment by jtobiska** [comment]
Thanks for making the changes so quickly!  I'm hoping to be able to spend some time tomorrow to confirm that it all works for our use cases.

**Comment by jtobiska** [comment]
I've done the first bit of testing, and it looks like it's mostly working.  The example that I provided in my original post is working as expected, which is fantastic!  Many thanks!!!

However I'm now seeing a couple cases that seem to be behaving strangely and I don't know if they're related:

**Case 1:**
```
>>> isinstance(conn.modules.__builtin__.StandardError(), conn.modules.__builtin__.StandardError)
True  # looks good
>>> isinstance(conn.modules.__builtin__.StandardError(), conn.modules.__builtin__.Exception)
False  # should be True since StandardError is a subclass of Exception
>>> isinstance(conn.modules.__builtin__.StandardError(), Exception)
True  # looks good
>>> isinstance(conn.modules.__builtin__.StandardError(), StandardError)
True  # looks good
```

**Case 2:**
```
>>> type(sys)
<type 'module'>  # base case
>>> type(conn.modules.sys)
<netref class 'rpyc.core.netref.__builtin__.module'>  # matches base case
>>> sys.__class__
<type 'module'>  # base case
>>> conn.modules.sys.__class__
<type 'module'>  # matches base case
>>> type(sys.__class__)
<type 'type'>  # base case
>>> type(conn.modules.sys.__class__)
<netref class 'rpyc.core.netref.__builtin__.module'>  # doesn't match.  Should be a netref class of "type" (or maybe just <type 'type'> itself?)
```

**Comment by jtobiska** [comment]
Just checking in on this since it's been a couple months since the last update.  Will it be possible to address the two cases I mentioned in my last comment?

**Comment by comrumino** [comment]
So, 4.1.3 was released just before writing this. Case 1 is fixed. Case 2, from [test_netref_hierachy.py](https://github.com/tomerfiliba/rpyc/blob/master/tests/test_netref_hierachy.py#L120-L122) (just noticed the typo...) the lines can be un-commented to show that the base cases switched in terms of passing.

Iirc, the test failures would keep switching. Once python2 is dropped, I intend to simplify the metaclass implementations---hopefully. Doing so could make it possible to fix Case 2 completely. The plan is to eventually take another crack at it after cleaning up compatibility complexities.

**Comment by jtobiska** [comment]
Sorry I've gone so long without giving a response since your last update!  Unfortunately I've been stuck on other priorities and haven't been able to devote any time until now.

I just tested out Case 1 above with v4.1.5 and I'm still seeing the same results as before:
```
>>> isinstance(conn.modules.__builtin__.StandardError(), conn.modules.__builtin__.Exception)
False  # should be True since StandardError is a subclass of Exception
```

So that one's not working, but Case 2 which I thought you said you weren't able to fix actually seems to be working perfectly.

**Comment by Jongy** [comment]
Found this issue while investigating a (possibly related) problem. All tests are performed on RPyC 5.0.1.

First of all, about the issue @jtobiska described in the last comment - I believe that's observed because the `Exception` & `BaseException` case goes through the `if other_id_pack[0] in netref.builtin_classes_cache:` check in `_handle_instancecheck`, while the `StandardError` case does not. I'm running Py3 (so not `StandardError`) but I can reproduce the same weird behavior with e.g `RuntimeError` -  it incorrectly yields `False`.

After I add `RuntimeError` to `netref.py:_builtin_types`, this finally yields:
```
>>> isinstance(conn.modules.builtins.RuntimeError(), conn.modules.builtins.Exception)
True
```

My second issue is with this discrepancy:
```
>>> conn = rpyc.classic.connect("127.0.0.1")
>>> c = conn.modules.collections.Counter()
>>> c.__class__
<class 'collections.Counter'>
>>> type(c)
<netref class 'rpyc.core.netref.type'>
>>> 
```

I'd like `type(c)` to return `rpyc.core.netref.Counter`.

I hope I'll have time further on this week, I'll try to solve these 2 issues (re @jtobiska 's issue, I'm not sure what's the correct approach, I don't think that adding all missing types to `_builtin_types` is the right approach...)

**Comment by comrumino** [comment]
> >>> isinstance(conn.modules.__builtin__.Exception(), conn.modules.__builtin__.BaseException)

I'll have to take a look into this. I think sometime ago I was fixing issues related to isinstance. 

Some more examples #390

**Comment by comrumino** [comment]
> I hope I'll have time further on this week, I'll try to solve these 2 issues (re @jtobiska 's issue, I'm not sure what's the correct approach, I don't think that adding all missing types to `_builtin_types` is the right approach...)

This seems like it could be an issue in how RPyC handles caching and iirc it does not cache the entire inheritance hierarchy. More over, if it fails to find it already cache it should try to locate it by other means which probably fails as well. 

Last time I worked on this it would fix some test and break others. There is probably logic errors in more than one place. I'll try to take another crack at this in the near future.

**Comment by comrumino** [comment]
@Jongy, any chance you want to take a look at my develop branch. Despite
```
----------------------------------------------------------------------
Ran 95 tests in 124.747s

FAILED (failures=2, errors=2, skipped=12)
```
The tests don't fail in the same way locally. Or anyone else, a second set of eyes would be make the fix move faster.
